### PR TITLE
ENH: Allow column grouping in DataFrame.plot

### LINKED ISSUE #29688: ENH: Group some columns in subplots with DataFrame.plot
`df.plot(subplots=True)`  will create one subplot per column. Is there a way to group multiple columns on the same subplot (and leave the rest of the column separated)?

I'd be happy to submit a PR if that's something you'd consider? In terms of API  `subplot` could accept a list of tuples where each tuple indicates which columns should be grouped together.

### PR DESCRIPTION
**Pull request by NicolasHug**
- [X] closes #29688
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This is an attempt at #29688 i.e. group some columns on the same ax when calling `df.plot`

See [this gist](https://gist.github.com/NicolasHug/4d1f60212ba69e6000982951ae47f203) for illustration of different supported plots

Thanks!

### CODE REVIEW

**Comment by WillAyd** [comment]
@NicolasHug is this still active? Looks like a merge conflict and CI failure need to be addressed

**Comment by NicolasHug** [comment]
I'm still happy to work on this. I just haven't received initial feedback yet (and don't want to push either, I understand this isn't necessarily a high priority)

> Before working more on this I'd appreciate some feedback to make sure this is going the in right direction, and that this is actually a desired feature.
Feel free to close otherwise, or LMK if you need to see more to decide :)

**Comment by WillAyd** [comment]
Understood. I doubt anyone has reviewed since CI has been red, so definitely want to get that green to start

**Comment by NicolasHug** [comment]
Will do. It's just linting though...

**Comment by WillAyd** [comment]
@charlesdong1991 any thoughts on this?

**Review by charlesdong1991 — CHANGES_REQUESTED** [review]
I've noticed this feature request a while ago and personally really like this feature and find it quite potentially useful ^^ @WillAyd  but I think the behaviour of this needs to be well defined.

Many thanks for the PR, it's very nice! @NicolasHug  some minor comments:
1. As mentioned in review, the input for `subplots` needs to be carefully defined, thus only checking if it is `Iterable` or `bool` might not be sufficient.
2. And it would be very nice if you could add some detailed docstring for `subplots` parameter because this is a behaviour change, i think somewhere in `PlotAccessor` in `_core.py`. 
3. And some more tests would be appreciated (see comments below)
4. Nicer to have a docstring for `_col_idx_to_axis_idx` although i think it is quite straightforward
4. Nicer to have a whatsnew note in `1.1.0` rst !

**Comment by NicolasHug** [comment]
Thanks a lot for the review @charlesdong1991 , I addressed your comments and added a bunch of validation / tests.

It looks like the code works for most kinds of plots. You can see a few illustrations in [this gist](https://gist.github.com/NicolasHug/4d1f60212ba69e6000982951ae47f203).

Looks like the subplot titles could be improved in some cases, I'll see what I can do. But otherwise I think this is ready for another round ;)

**Comment by NicolasHug** [comment]
Regarding the docstring, I'm a bit confused because it seems that `subplots` have been removed from most docstrings? i.e. `git grep "subplots :"` only gives one hit, for `boxplot_frame_groupby()`. I can't find it for `PlotAccessor`.

**Review by charlesdong1991 — COMMENTED** [review]
very nice! some minor comments

Could you also please make some docs for `subplots`? I think you could add it in `PlotAcessor`, there are some missing docs for parameters, so very nice this one is found out! ^^

File: pandas/plotting/_matplotlib/core.py
Comment (NicolasHug) [reply pandas/plotting/_matplotlib/core.py:140]: The thing is, we want `group` to be an iterable but we don't want `group` to be a str, which is also iterable.

Only using `if any(not isinstance(group, Iterable)` would allow strings and we don't want that

File: pandas/plotting/_matplotlib/core.py
Comment (NicolasHug) [reply pandas/plotting/_matplotlib/core.py:131]: I agree it doesn't look good, but it seems to be consistent (i.e. equally bad) as when `subplots=True`. I updated the [notebook](https://gist.github.com/NicolasHug/4d1f60212ba69e6000982951ae47f203) at the end where you can see how it looks. So maybe it can be kept? No strong opinion!

File: doc/source/whatsnew/v1.1.0.rst
Comment (charlesdong1991) [inline doc/source/whatsnew/v1.1.0.rst:103]: we usually put issue number here, not the PR number, so in this case, would use `29688`

File: pandas/plotting/_matplotlib/core.py
Comment (charlesdong1991) [inline pandas/plotting/_matplotlib/core.py:118]: Pandas started to add annotations for the new change a while ago, and now we would prefer to have it annotated especially given this will be a significant change to `subplots`, so could you please add annotations above in init?

Comment (charlesdong1991) [reply pandas/plotting/_matplotlib/core.py:118]: you might need to also add some annotations in your code to make mypy happy, if it is the case, feel free to add

File: pandas/plotting/_matplotlib/core.py
Comment (charlesdong1991) [inline pandas/plotting/_matplotlib/core.py:751]: the annotation also is needed for the new method, thanks

**Review by charlesdong1991 — CHANGES_REQUESTED** [review]
Hey, thanks very much for your work, let's give it another push! 

could you please rebase and resolve conflicts, having some annotations will be good!

File: pandas/plotting/_matplotlib/core.py
Comment (charlesdong1991) [reply pandas/plotting/_matplotlib/core.py:129]: i have no strong opinion either, my only concern is for pie chart, if they share the same category, they will have the same color in the pie, but different sizes depending on the corresponding values. 

but no strong opinion

**Comment by jreback** [comment]
@NicolasHug can you merge master and update to comments

**Comment by NicolasHug** [comment]
sorry I'm short on time ATM and I won't be able to come back to this before a few weeks

**Comment by charlesdong1991** [comment]
@nicholsn thanks for this nice proposal, it definitely will be nice to have!!

and I am very sorry for taking so long for the feedbacks!

not sure if I am allowed to finish this PR for you to make sure you have the credit as well.  I will take a look and see!

**Comment by NicolasHug** [comment]
no worries at all, I know what it's like ;)

I definitely plan to come back to it later, but if you'd rather have this in sooner then feel free to take over!

**Comment by charlesdong1991** [comment]
ahh, great!! then will leave it on you, and wish to see you coming back to implement this nice feature! ^^  @NicolasHug

File: pandas/plotting/_matplotlib/core.py
Comment (NicolasHug) [inline pandas/plotting/_matplotlib/core.py:164]: I would rather just do `cols_in_groups = set(cols_in_groups)` but I couldn't make pypy happy otherwise

File: pandas/plotting/_matplotlib/core.py
Comment (NicolasHug) [inline pandas/plotting/_matplotlib/core.py:127]: for some reason I have to declare the type here once again despite having specified it already above.

Also note I have to add `int` so that mypy doesn't complain, but users should only pass strings (column names) instead of ints (indices)

Comment (simonjayhawkins) [reply pandas/plotting/_matplotlib/core.py:127]: ci failing with Check for use of comment-based annotation syntax. please use py3.6+ syntax https://mypy.readthedocs.io/en/stable/type_inference_and_annotations.html?highlight=variable%20annotation#explicit-types-for-variables

> for some reason I have to declare the type here once again despite having specified it already above.

from https://mypy.readthedocs.io/en/stable/type_inference_and_annotations.html#type-inference

> Mypy considers the initial assignment as the definition of a variable.

self.subplots is hence infered to be bool without the variable type annotation. (and the mypy error then makes sense `error: Incompatible types in assignment (expression has type "List[<nothing>]", variable has type "bool")`

could use an alias to reduce duplication, and/or could declare variable type in class scope to reduce noise.

also maybe Sequence[Sequence[Union[str, int]]]]] -> Sequence[Sequence[Label]]]]

File: pandas/plotting/_matplotlib/core.py
Comment (simonjayhawkins) [reply pandas/plotting/_matplotlib/core.py:164]: Again, if you want to reuse variable names for different types, needs a variable annotation before first assignment. I think OK to use different name.

File: pandas/plotting/_matplotlib/core.py
Comment (simonjayhawkins) [inline pandas/plotting/_matplotlib/core.py:88]: using `list of iterables` in docstring. mostly we want these consistent.

Comment (simonjayhawkins) [reply pandas/plotting/_matplotlib/core.py:88]: don't need the Optional

File: pandas/plotting/_matplotlib/core.py
Comment (simonjayhawkins) [reply pandas/plotting/_matplotlib/core.py:127]: I think that this is adding quite a bit of code to \_\_init__. maybe create a _validate_subplot_kwarg method (the variable annotation issue could disappear (with function parameters and return type typed)

File: pandas/plotting/_matplotlib/core.py
Comment (simonjayhawkins) [inline pandas/plotting/_matplotlib/core.py:156]: do we need the counter, can you compare the length of a set with the len of list

Comment (NicolasHug) [reply pandas/plotting/_matplotlib/core.py:156]: Using a counter allows printing the duplicates. let me know what you prefer

Comment (simonjayhawkins) [reply pandas/plotting/_matplotlib/core.py]: fair enough, if that's a requirement.

**Comment by NicolasHug** [comment]
thanks for the review and for the tips @simonjayhawkins , I have updated the PR accordingly.
The test failure seems to be related to a networking issue (unclosed socket), I'm not sure this is directly related to this PR?

**Comment by NicolasHug** [comment]
Looks like the CI failure was magically resolved. I think this is ready for another round! thanks for the reviews so far.

**Review by jreback — CHANGES_REQUESTED** [review]
can you show a sample plot with this api. are there other ways to do this, e. g a for loop?

**Comment by charlesdong1991** [comment]
> can you show a sample plot with this api.

@NicolasHug indeed made a nice sample ^^, see in PR description on top: 
> See this gist for illustration of different supported plots

the link is https://gist.github.com/NicolasHug/4d1f60212ba69e6000982951ae47f203  
there was a discussion about pie plots since they don't as good as others, https://github.com/pandas-dev/pandas/pull/29944#discussion_r404364673 @jreback your opinion is very welcomed!

**Comment by NicolasHug** [comment]
Thanks @jreback and @charlesdong1991 . All of your requested changes were addressed.

**Comment by NicolasHug** [comment]
> are there other ways to do this, e. g a for loop?

I know of no other way in pandas. Like any other supported plot in pandas, this is of course possible with pure matplotlib code, but with more boilerplate.

**Comment by NicolasHug** [comment]
I'm going to stop addressing merge conflicts because they seem to occur 5 minutes after every single commit that I make. They're only about conflicts in the imports. I hope this doesn't hurt reviewing.

**Comment by charlesdong1991** [comment]
thanks @NicolasHug , IMHO, this is a very nice PR and is very close to be merged, many thanks for your patience!

i think the conflict won't affect your PR, so it's okay to resolve the conflict once @jreback comments on https://github.com/pandas-dev/pandas/pull/29944#issuecomment-687559969 and there is kinda consensus on how to address.

One other thing worth mentioning is https://github.com/pandas-dev/pandas/pull/29944#discussion_r483902388 this PR should be merged to  `Plotting` section in `1.2` instead of `1.1.2`, since `1.1.2` will be mainly focused on regression related issues.

**Comment by NicolasHug** [comment]
i'm still happy to address any future review

**Comment by arw2019** [comment]
@NicolasHug thanks for being patient! Can you merge master one more time and resolve conflicts?

cc @jreback  xref https://github.com/pandas-dev/pandas/pull/29944#issuecomment-687559969

**Comment by arw2019** [comment]
cc @jreback @charlesdong1991 this looks close to going in in case we want it for 1.2

**Review by arw2019 — COMMENTED** [review]
@NicolasHug sorry about the delay here! Could you merge master again (there are some conflicts) and move the whatsnew to 1.3

**Comment by arw2019** [comment]
ping @jreback @charlesdong1991

File: doc/source/whatsnew/v1.2.0.rst
Comment (jreback) [inline doc/source/whatsnew/v1.2.0.rst:215]: move to 1.3

File: pandas/plotting/_core.py
Comment (jreback) [inline pandas/plotting/_core.py:637]: add versionupdated 1.3

File: pandas/plotting/_matplotlib/core.py
Comment (jreback) [inline pandas/plotting/_matplotlib/core.py:214]: can you give a doc-string that explains what this is doing

e.g. Parameters / Returns / Raises sections

File: pandas/plotting/_matplotlib/core.py
Comment (jreback) [inline pandas/plotting/_matplotlib/core.py:241]: test hitting this?

File: pandas/plotting/_matplotlib/core.py
Comment (jreback) [inline pandas/plotting/_matplotlib/core.py:233]: test hitting this?

File: pandas/plotting/_matplotlib/core.py
Comment (jreback) [inline pandas/plotting/_matplotlib/core.py:219]: test hitting these cases?

File: pandas/plotting/_matplotlib/core.py
Comment (jreback) [inline pandas/plotting/_matplotlib/core.py:250]: test hitting this?

File: pandas/plotting/_matplotlib/core.py
Comment (jreback) [inline pandas/plotting/_matplotlib/core.py:213]: the return signature is key here Union[bool, List[Tuples[int]]]

**Review by jreback — CHANGES_REQUESTED** [review]
don't we *already* do this when we `df.groupb(...).plot.line()` for example?

File: pandas/plotting/_matplotlib/core.py
Comment (NicolasHug) [reply pandas/plotting/_matplotlib/core.py:240]: This is already tested

**Comment by NicolasHug** [comment]
Thank you for the review @jreback, comments were addressed.

> don't we already do this when we df.groupb(...).plot.line() for example?

I'm probably missing something but `groupby.plot` does something very different from the proposed PR: `groupby.plot` will group by **value** while the proposed PR simply groups **entire columns** within the same subplots.

**Comment by NicolasHug** [comment]
CI Failures are unrelated I think

**Comment by simonjayhawkins** [comment]
> CI Failures are unrelated I think

```
Performing static analysis using mypy
pandas/plotting/_matplotlib/core.py:304: error: Incompatible return value type (got "List[Tuple[int, ...]]", expected "Union[bool, List[Tuple[int]]]")  [return-value]
Found 1 error in 1 file (checked 1177 source files)
Performing static analysis using mypy DONE
```

mypy needs to be green before we can consider merging.

**Comment by NicolasHug** [comment]
Thanks, fixed mypy. Should `Linux py38_np_dev` still fail, I believe it's still unrelated to the proposed changes.

**Comment by lithomas1** [comment]
@NicolasHug Mind merging master/resolving conflicts? This would be quite nice to get into 1.3 I think.

File: pandas/plotting/_matplotlib/core.py
Comment (charlesdong1991) [inline pandas/plotting/_matplotlib/core.py:108]: I _think_ in general, pandas uses `Union` instead of `|`, would be nicer to align with other similar cases in the codebase.

**Review by charlesdong1991 — APPROVED** [review]
Hi, sorry for the wait. I like this PR and the feature, and will be a nice one to have in 1.3! Just a minor comment, otherwise look very good to me.

File: pandas/plotting/_matplotlib/core.py
Comment (charlesdong1991) [reply pandas/plotting/_matplotlib/core.py:108]: ah, okay! sorry about that, i was not aware of such update in pre-commit hook.

Comment (charlesdong1991) [reply pandas/plotting/_matplotlib/core.py:108]: emm, interesting, still see `Union` used in codebase. maybe @simonjayhawkins or @MarcoGorelli knows better about this change in the precommit hook?

**Review by lithomas1 — CHANGES_REQUESTED** [review]

File: pandas/plotting/_matplotlib/core.py
Comment (lithomas1) [inline pandas/plotting/_matplotlib/core.py:229]: Looks like mypy is failing here.
From https://github.com/pandas-dev/pandas/pull/29944/checks?check_run_id=2548207329
```
pandas/plotting/_matplotlib/core.py:229: error: Bracketed expression "[...]" is not valid as a type  [misc]
pandas/plotting/_matplotlib/core.py:229: note: Did you mean "List[...]"?
```
I think this should be ``list[tuple[int, ...]]``

Comment (MarcoGorelli) [reply pandas/plotting/_matplotlib/core.py]: Hi @NicolasHug - are you sure it was removed by a pre-commit job? I just tried:
```
git checkout -b pr/NicolasHug-subplot_groups master
git fetch https://github.com/NicolasHug/pandas.git subplot_groups
git reset --hard FETCH_HEAD
git checkout 827c5b5
pre-commit run --files pandas/plotting/_matplotlib/core.py
```
and all the checks passed. Indeed, I don't know of any hook which would split an annotation like that in `1517ae3` - perhaps it was an error while fixing a merge conflict? Else, do you happen to remember what you ran? Thanks

File: pandas/tests/plotting/frame/test_frame.py
Comment (charlesdong1991) [inline pandas/tests/plotting/frame/test_frame.py:2134]: emm, are these 2 tests by mistake added from other PRs?

**Review by charlesdong1991 — APPROVED** [review]
It looks good to me except for a comment about tests from other pr?

In addition to @jreback , maybe @MarcoGorelli or @ivanovmg could also have time to take a look?

File: pandas/plotting/_core.py
Comment (ivanovmg) [inline pandas/plotting/_core.py:655]: Just to clarify, are the docs rendered correctly with this empty line?

File: pandas/plotting/_matplotlib/core.py
Comment (ivanovmg) [inline pandas/plotting/_matplotlib/core.py:267]: What if ``subplots`` is boolean, then would it be necessary to validate ``self._kind``?
If so, then, I guess, it deserves a separate method to be invoked at the moment of ``self.kind`` initialization.

By the way, this is quite confusing for me (probably regardless of the given PR): inside ``MPLPlot`` one initializes ``self.kind``, but it seems that it is not used anywhere, only ``self._kind`` does.
Am I missing something?

File: pandas/plotting/_matplotlib/core.py
Comment (ivanovmg) [inline pandas/plotting/_matplotlib/core.py:838]: Can you type a return value?

File: pandas/plotting/_matplotlib/core.py
Comment (ivanovmg) [inline pandas/plotting/_matplotlib/core.py:845]: Is this ``else`` clause required? I guess, that it may be dropped.

File: pandas/tests/plotting/frame/test_frame.py
Comment (ivanovmg) [inline pandas/tests/plotting/frame/test_frame.py:2075]: Good parametrization!

File: pandas/plotting/_matplotlib/core.py
Comment (ivanovmg) [inline pandas/plotting/_matplotlib/core.py:284]: Is it really a desirable behavior?
Why not allowing one to plot the same line against different sets?

File: pandas/tests/plotting/frame/test_frame.py
Comment (ivanovmg) [inline pandas/tests/plotting/frame/test_frame.py:2096]: Is it really a desirable behavior (see other comment).

**Review by ivanovmg — COMMENTED** [review]
It seems like a reasonable extension of ``subplots`` functionality!
I have one concern regarding the inability to plot the same column on two different subplots (reflected in the comments).
Otherwise looks good!

File: pandas/plotting/_matplotlib/core.py
Comment (NicolasHug) [reply pandas/plotting/_matplotlib/core.py:284]: Currently (in master), only one column per plot is possible.
Allowing more than one column per plot should be possible, but would make the PR more complex. This PR il already 1.5+ years long so I would leave that improvement for a subsequent PR.

File: pandas/plotting/_core.py
Comment (NicolasHug) [reply pandas/plotting/_core.py:655]: We need an empty line because we're starting a list.
Is there a CI job that builds the docs? I couldn't find it

File: pandas/plotting/_matplotlib/core.py
Comment (NicolasHug) [reply pandas/plotting/_matplotlib/core.py:267]: > What if subplots is boolean, then would it be necessary to validate self._kind?

When subplots is boolean we return immediately as we fall back to the current behaviour, as in master. I haven't double checked but I would hope that `self._kind` is validated separately - I mean, it's `_kind`, not `kind`.

The validation happening in `_validate_subplots_kwarg` is only for `subplots`, and its interactions with other parameters (such as `_kind`) when `subplots` is an iterable. But its job is not to validate `_kind`.

File: pandas/plotting/_core.py
Comment (ivanovmg) [reply pandas/plotting/_core.py:655]: Yes, CI runs to build the docs.
But you can build the particular page (for ``DataFrame.plot``) yourself to ensure it is correct.
https://pandas.pydata.org/pandas-docs/dev/development/contributing_documentation.html#building-the-documentation

File: pandas/plotting/_matplotlib/core.py
Comment (charlesdong1991) [reply pandas/plotting/_matplotlib/core.py:284]: yeah, agree, could be a follow-up enhancement PR!

File: pandas/plotting/_core.py
Comment (NicolasHug) [reply pandas/plotting/_core.py:656]: I couldn't build the docs

```
python make.py clean
python make.py --single pandas.DataFrame.plot
```

fails with stuff like `WARNING: autodoc: failed to import method 'plot.scatter' from module 'DataFrame'; the following exception was raised:
No module named 'DataFrame'`

So instead I just added a line so that the docstring for subplots is consistent with the rest. If the rest is fine, then this one should be fine too.

**Review by ivanovmg — APPROVED** [review]
Looks good to me.
Pending CI checks.

File: pandas/plotting/_matplotlib/core.py
Comment (ivanovmg) [reply pandas/plotting/_matplotlib/core.py:845]: I suppose, that once you added return type, ``mypy`` started complaining.
Inside the loop you return only in case ``if col_idx in group``, this is the reason for the ``mypy`` error.
I suggest that you modify it:
```
for group_idx, group in enumerate(self.subplots):
    if col_idx in group:
    return group_idx
raise KeyError(f"Failed to find {col_idx} in {self.subplots}")
```
Or something like this.

**Review by ivanovmg — CHANGES_REQUESTED** [review]

File: pandas/plotting/_matplotlib/core.py
Comment (NicolasHug) [reply pandas/plotting/_matplotlib/core.py:847]: I think mypy is wrong here. Because of the validation done earlier, there *must* be a valid column. Raising an exception might make mypy happy but it would never be hit (because of the validation), and we wouldn't be able to test against it. I changed the for loop into a `next()`, which seems to trick mypy into being reasonable again.

**Comment by jreback** [comment]
@NicolasHug was looking good. if you want to continue, pls merge master and can relook.

**Comment by jreback** [comment]
sorry for the delay here

@NicolasHug was looking good. if you want to continue, pls merge master and can relook.

File: doc/source/whatsnew/v1.3.0.rst
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:262]: can you move to 1.5

**Comment by NicolasHug** [comment]
PR is up to date with `main` @jreback

**Comment by jreback** [comment]
@NicolasHug can you merge master and we'll see if can get in

**Review by mroeschke — APPROVED** [review]
Thanks for sticking with this @NicolasHug. I merged master and did a small refactor to validate the subplot groups in one pass.

I think this looks pretty good to merge. I believe it requires @jreback approval to move forward.

**Review by jreback — APPROVED** [review]

**Pull request merged**
